spec/safe_statements_spec.rb

```diff
@@ -538,17 +538,71 @@ def up
             end.to raise_error PgHaMigrations::UnsafeMigrationError
           end
 
-          it "add column of default is not set" do
-            migration = Class.new(migration_klass) do
-              def up
-                unsafe_create_table :foos
-                safe_add_column :foos, :bar, :text
+          context 'with Postgres 11+' do
+            before do
+              allow(ActiveRecord::Base.connection).to receive(:postgresql_version).and_return(11_00_00)
+            end
+
+            it "allows setting a default" do
+              migration = Class.new(migration_klass) do
+                def up
+                  unsafe_create_table :foos
+                  safe_add_column :foos, :bar, :text, :default => "baz"
+                end
               end
+              
+              migration.suppress_messages { migration.migrate(:up) }
+
+              expect(ActiveRecord::Base.connection.columns("foos").detect { |column| column.name == "bar" }.default).to eq("baz")
+
+              ActiveRecord::Base.connection.execute("INSERT INTO foos SELECT FROM (VALUES (1)) t")
+              expect(ActiveRecord::Base.connection.select_value("SELECT bar FROM foos")).to eq("baz")
             end
 
-            migration.suppress_messages { migration.migrate(:up) }
+            it "allows setting null => false" do
```

Comment:
This still isn't safe with PG11+ right? Adding a `NOT NULL` constraint still requires a full table scan.

Reply:
Perhaps you were thinking of the feature to convert a `CHECK col IS NOT NULL` constraint into a `NOT NULL` constraint?

Reply:
By the way, I have a branch to add more generic constraints, which would lay he groundwork for a function to use the new PG functionality I referenced.

Reply:
It's not safe for change column, but it's not possible for add column without a default value (because the table is queried for null values)

Reply:
Oh, I didn't really realize what you were intending at first, and that explanation helped.

In general we're trying to achieve no surprises when a migration runs in a production environment, so I'd tend to consider `safe_add_column :foos, :bar, :text, :null => false` to be unsafe in the sense that it can't be known to be executable (since there could be rows in the table already, which would at least fail _fast_ but still fail).

In our internal process for that case we'd use the `unsafe_add_column` variant and also require that someone from the team owning the table document the row counts in all deployed environments and sign off that the code can't currently write any new rows.